sway/config/output.c

```diff
@@ -257,24 +257,43 @@ static void set_mode(struct wlr_output *output, struct wlr_output_state *pending
 		return;
 	}
 
-	struct wlr_output_mode *mode, *best = NULL;
+	struct wlr_output_mode *mode, *best = NULL, *highest = NULL;
+	int best_diff_mhz = INT_MAX;
 	wl_list_for_each(mode, &output->modes, link) {
 		if (mode->width == width && mode->height == height) {
-			if (mode->refresh == mhz) {
-				best = mode;
-				break;
+			// If a target refresh rate is given, match it as closely as possible
+			if (mhz >= 0) {
+				int diff_mhz = abs(mode->refresh - mhz);
+				if (diff_mhz < best_diff_mhz) {
+					best_diff_mhz = diff_mhz;
+					best = mode;
+					if (best_diff_mhz == 0) {
+						break;
+					}
+				}
 			}
-			if (best == NULL || mode->refresh > best->refresh) {
-				best = mode;
+			if (highest == NULL || mode->refresh > highest->refresh) {
```

Comment:
I wonder if we should do away with highest. Even if we're more than 1Hz away, picking the closest refresh rate to what the user specified is probably fine. When no refresh rate is specified, we could just set `mhz = INT_MAX`, in which the search for smallest diff also finds the largest value.

Reply:
Sounds like a good solution to me, will implement it like this.

The only downside I see is that e.g. specifying `90Hz`, when only ~`60Hz` and ~`120Hz` are available, makes it less predictable which will be chosen.
The limit ensures that either around the user-specified or the highest refresh rate will be applied.